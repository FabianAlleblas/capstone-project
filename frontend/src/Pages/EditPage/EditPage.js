import PropTypes from 'prop-types'
import { useHistory, useLocation } from 'react-router-dom'
import styled from 'styled-components/macro'
import EditPlantForm from '../../Components/Forms/EditPlantForm'
import Header from '../../Components/Header/Header'
import ErrorModal from '../../Components/Modals/ErrorModal'

EditPage.propTypes = {
  updatePlantData: PropTypes.func.isRequired,
  plantList: PropTypes.array.isRequired,
  deletePlantData: PropTypes.func.isRequired,
}

export default function EditPage({
  updatePlantData,
  plantList,
  deletePlantData,
}) {
  const history = useHistory()
  const query = useQuery()
  const plant = plantById(useQuery)

  if (!plant) {
    return (
      <ErrorModalFixed closeErrorModal={() => history.push('/')}>
        Plant not found!
      </ErrorModalFixed>
    )
  }

  return (
    <>
      <FixedHeader showDeleteButton secondaryStyle onClickDelete={handleDelete}>
        Edit Your Plant
      </FixedHeader>
      <FormContainer>
        <EditPlantForm
          updatePlantData={updatePlantData}
          deletePlantData={deletePlantData}
          plant={plant}
        />
      </FormContainer>
    </>
  )

  function useQuery() {
    return new URLSearchParams(useLocation().search)
  }

  function plantById() {
    const plantId = parseInt(query.get('id'))
    const plant = plantList.find((plant) => plant.id === plantId)

    return plant
  }

  function handleDelete() {
    deletePlantData(plant.id)
    history.push('/')
  }
}

const FixedHeader = styled(Header)`
  left: 0;
  position: fixed;
  top: 0;
`

const FormContainer = styled.main`
  align-items: center;
  background-color: var(--primary-dark);
  display: flex;
  height: 100vh;
  padding: 40px 40px 0;
  place-content: center;
`

const ErrorModalFixed = styled(ErrorModal)`
  height: 100vh;
  left: 0;
  position: absolute;
  top: 0;
`
