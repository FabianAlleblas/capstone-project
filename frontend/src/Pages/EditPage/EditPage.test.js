import { render } from '@testing-library/react'
import user from '@testing-library/user-event'
import 'jest-styled-components'
import { MemoryRouter } from 'react-router-dom'
import EditPage from './EditPage'
import defaultPlantImage from '../../assets/plant-images/default_plant.jpg'

const plantList = [
  {
    id: 2,
    name: 'Carl',
    species: 'Monstera deliciosa',
    specialInfo: 'variegated',
  },
]

const mockHistoryPush = jest.fn()

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}))

const updatePlandDataMock = jest.fn()
const deletePlantDataMock = jest.fn()

describe('EditPage', () => {
  it('renders correctly', () => {
    const { container } = render(
      <MemoryRouter initialEntries={['/edit-page?id=2']}>
        <EditPage
          updatePlantData={updatePlandDataMock}
          deletePlantData={deletePlantDataMock}
          plantList={plantList}
        />
      </MemoryRouter>
    )

    expect(container).toMatchSnapshot()
  })

  it('renders with the correct data', () => {
    const { getByLabelText, getByTestId } = render(
      <MemoryRouter initialEntries={['/edit-page?id=2']}>
        <EditPage
          updatePlantData={updatePlandDataMock}
          deletePlantData={deletePlantDataMock}
          plantList={plantList}
        />
      </MemoryRouter>
    )

    expect(getByLabelText(/Your plants name/i)).toHaveValue(plantList[0].name)
    expect(getByLabelText(/The species of your plant/i)).toHaveValue(
      plantList[0].species
    )
    expect(getByLabelText(/Special infos/i)).toHaveValue(
      plantList[0].specialInfo
    )
    expect(getByTestId('img-input-wrapper')).toHaveStyle(
      `background-image: url(${defaultPlantImage})`
    )
  })

  it('calls deletePlantData with the correct data by clicking the delete button and call history push', () => {
    const { getByRole } = render(
      <MemoryRouter initialEntries={['/edit-page?id=2']}>
        <EditPage
          updatePlantData={updatePlandDataMock}
          deletePlantData={deletePlantDataMock}
          plantList={plantList}
        />
      </MemoryRouter>
    )
    user.click(getByRole('button', { name: /deleteicon/i }))
    expect(deletePlantDataMock).toHaveBeenCalledWith(plantList[0].id)
    expect(mockHistoryPush).toHaveBeenCalled()
  })

  it('renders the error modal when not finding the plant id', () => {
    const { getByText } = render(
      <MemoryRouter initialEntries={['/edit-page?id=3']}>
        <EditPage
          updatePlantData={updatePlandDataMock}
          deletePlantData={deletePlantDataMock}
          plantList={plantList}
        />
      </MemoryRouter>
    )
    expect(getByText(/plant not found/i)).toBeInTheDocument()
  })
})
