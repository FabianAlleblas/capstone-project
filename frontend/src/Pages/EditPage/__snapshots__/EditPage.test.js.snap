// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditPage renders correctly 1`] = `
.c13 {
  background-color: var(--primary-light);
  border-radius: 22px;
  border: none;
  box-shadow: 0 0 4px var(--shadow);
  color: var(--primary-dark);
  font-family: 'Josefin sans';
  font-size: 1.25rem;
  font-weight: 700;
  padding: 12px 16px 8px;
  width: 100%;
}

.c15 {
  background-color: var(--primary-dark);
  border-radius: 22px;
  border: 4px solid var(--primary-light);
  box-shadow: 0 0 4px var(--shadow);
  color: var(--primary-light);
  font-family: 'Josefin sans';
  font-size: 1.25rem;
  font-weight: 700;
  padding: 8px 12px 4px;
  width: 100%;
}

.c7 {
  background-color: var(--primary-dark);
  display: grid;
  gap: 20px;
}

.c8 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-image: url(default_plant.jpg);
  background-position: center;
  background-size: cover,contain;
  border-radius: 50%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 148px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin: 0 auto 10px;
  position: relative;
  width: 148px;
}

.c8 svg {
  margin: 0 4px 4px 0;
}

.c9 {
  height: 0.1px;
  opacity: 0;
  overflow: hidden;
  position: absolute;
  width: 0.1px;
  z-index: -1;
}

.c10 {
  color: var(--form-font-color);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  font-size: 1.2rem;
}

.c11 {
  border-radius: 22px;
  border: none;
  color: var(--form-font-color);
  display: block;
  font-family: 'Josefin sans',sans-serif;
  font-size: 1rem;
  margin-top: 8px;
  padding: 14px 20px 10px;
  width: 100%;
}

.c11:focus {
  box-shadow: 0 0 4px var(--shadow) inset;
  outline: none;
}

.c12 {
  display: grid;
  gap: 20px;
  grid-template-columns: 1fr 1fr;
  padding: 30px 0 0;
}

.c14 {
  opacity: 1;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
  z-index: 10;
}

.c2 {
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  background: var(--primary-dark);
  border-radius: 0 0 50px 0;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  padding: 20px 30px 13px 20px;
}

.c3 {
  color: var(--primary-light);
  font-size: 1.75rem;
  font-weight: 700;
  line-height: 1;
}

.c5 {
  margin-right: 12px;
}

.c4 {
  background-color: transparent;
  border: none;
}

.c1 {
  left: 0;
  position: fixed;
  top: 0;
}

.c6 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: var(--primary-dark);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100vh;
  padding: 40px 40px 0;
  place-content: center;
}

<div>
  <header
    class="c0 c1"
  >
    <div
      class="c2"
      data-testid="heading-container"
    >
      <h1
        class="c3"
      >
        Edit Your Plant
      </h1>
    </div>
    <button
      class="c4"
    >
      <svg
        class="c5"
      >
        deleteicon.svg
      </svg>
    </button>
  </header>
  <main
    class="c6"
  >
    <form
      class="c7"
    >
      <label
        class="c8"
        data-testid="img-input-wrapper"
        src="default_plant.jpg"
      >
        <input
          accept="image/png, image/jpeg"
          alt="image-input"
          class="c9"
          name="image"
          type="file"
        />
        <svg>
          addimgicon.svg
        </svg>
      </label>
      <label
        class="c10"
      >
        Your plants name*:
        <input
          class="c11"
          maxlength="28"
          name="name"
          required=""
          type="text"
          value="Carl"
        />
      </label>
      <label
        class="c10"
      >
        The species of your plant*:
        <input
          class="c11"
          maxlength="28"
          name="species"
          required=""
          type="text"
          value="Monstera deliciosa"
        />
      </label>
      <label
        class="c10"
      >
        Special infos:
        <input
          class="c11"
          maxlength="28"
          name="specialInfo"
          type="text"
          value="variegated"
        />
      </label>
      <div
        class="c12"
      >
        <button
          class="c13 c14"
        >
          Update
        </button>
        <button
          class="c15"
          type="button"
        >
          Cancel
        </button>
      </div>
    </form>
  </main>
</div>
`;
